protocol_rfcs/variant-type.md

```diff
@@ -0,0 +1,98 @@
+# Variant Data Type
+**Associated Github issue for discussions: https://github.com/delta-io/delta/issues/2864**
+
+This protocol change adds support for the Variant data type.
+The Variant data type is beneficial for storing and processing semi-structure data.
+
+--------
+
+> ***New Section after the [Clustered Table](#clustered-table) section***
+
+# Variant Data Type
+
+This feature introduces the Variant data type, for processing with semi-structured data.
+The schema serialization method is described in [Schema Serialization Format](#schema-serialization-format).
+
+To support this feature:
+- The table must be on Reader Version 3 and Writer Version 7
+- The feature `variantType` must exist in the table `protocol`'s `readerFeatures` and `writerFeatures`, either during its creation or at a later stage.
+
+## Example JSON-Encoded Delta Table Schema with Variant types
+
+```
+{
+  "type" : "struct",
+  "fields" : [ {
+    "name" : "raw_data",
+    "type" : "variant",
+    "nullable" : true,
+    "metadata" : { }
+  }, {
+    "name" : "variant_array",
+    "type" : {
+      "type" : "array",
+      "elementType" : {
+        "type" : "variant"
+      },
+      "containsNull" : false
+    },
+    "nullable" : false,
+    "metadata" : { }
+  } ]
+}
+```
+
+## Variant data in Parquet
+
+The Variant data type is represented as 2 binary encoded values, according to the [Variant binary encoding specification](https://github.com/apache/spark/blob/master/common/variant/README.md).
+The 2 binary values are named `value` and `metadata`.
+
+When writing Variant data to parquet files, the Variant data is written as a single Parquet struct, with the following fields:
+
+Struct field name | Parquet primitive type | Description
+-|-|-
+value | binary | The binary-encoded Variant value, as described in [Variant binary encoding](https://github.com/apache/spark/blob/master/common/variant/README.md)
+metadata | binary | The binary-encoded Variant  metadata, as described in [Variant binary encoding](https://github.com/apache/spark/blob/master/common/variant/README.md)
+
+The parquet struct must include the 2 struct fields `value` and `metadata`.
+Supported writers must write the 2 binary fields, and supported readers must read the 2 binary fields.
+Struct fields which start with `_` (underscore) can be safely ignored.
+The only non-ignorable fields must be `value` and `metadata`.
```

Comment:
This seems to say (somewhat implicitly) that any field that is not `value` or `metadata` should start with an underscore? I don't see why that is necessary, if we define that only `value` and `metadata` should be non-ignorable.

Reply:
Something we would like to support in the future is to introduce another non-ignorable struct field. That would require a new table feature, but it would depend on this table feature. In that case, how can we specify in the protocol that the new table feature would use an additional, non-ignorable field?

Reply:
The Delta spec already states that clients should ignore unrecognized fields, does that not suffice?
If the idea is to carve out a "safe" namespace for custom fields, then `_` prefix seems reasonable?